### ISSUE TITLE
Effect link exception

### ISSUE THREAD

**Issue by Zoldike** [MonoGame/MonoGame#3639]
#### The situation

We are using custom SpriteBatch effects in our game. For example:

```
sampler TextureSampler : register(s0);
uniform float2 Offset = float2(0, 0);

float4 main(float4 position : SV_Position, float4 color : COLOR0, float2 texCoord : TEXCOORD0) : COLOR0
{
    return tex2D(TextureSampler, texCoord + Offset);
}

technique Displacement
{
    pass Pass1
    {
#if HLSL
        PixelShader = compile ps_4_0_level_9_3 main();
#else
        PixelShader = compile ps_2_0 main();
#endif
    }
}
```
#### The exception

"No vertex shader attached."
https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Graphics/Shader/ShaderProgramCache.cs#L138

It happenes **rarely** and **only on Android devices** (WP8, W8, iOS are fine).

> Hi, it happenned only once, and I never had this problem again.

Do we need to directly specify a pass through vertex shader? Any idea?

**Comment by tomspilman** [comment]
> "No vertex shader attached."

Is that your error or is it the "Unable to link effect program" exception you linked to?

What does your `SpriteBatch` code look like that uses this effect?

> Do we need to directly specify a pass through vertex shader?

No.

When you don't specify a vertex shader to `SpriteBatch` it will use the normal sprite vertex shader.  How you got it to not set one at all is the mystery here.

**Comment by Zoldike** [comment]
> Is that your error or is it the "Unable to link effect program" exception you linked to?

I have made a small change in the exception message to get the error log, nothing else.

Basicly we are using 4 custom SpriteBatch effects, three of them are post process shaders.
```
if (selected != null)
    this.GraphicsDevice.SetRenderTarget(this.renderTarget);

base.Draw(gameTime);

if (selected != null)
{
    this.GraphicsDevice.SetRenderTarget(null);
    this.GraphicsDevice.Clear(Color.White);
    this.ScreenManager.SpriteBatch.Begin(SpriteSortMode.Deferred, selected.BlendState, SamplerState.LinearClamp, DepthStencilState.None, RasterizerState.CullNone, selected.Effect);
    this.ScreenManager.SpriteBatch.Draw(renderTarget, Vector2.Zero, Color.White);
    this.ScreenManager.SpriteBatch.End();
}
```

There is only one element in the visual tree that must be rendered using a custom effect. So we have to break the Begin-End segment once per frame.
```
 if (this.Effect != null)
{
    this.SpriteBatch.End();
    this.SpriteBatch.Begin(SpriteSortMode.Deferred, BlendState.AlphaBlend, SamplerState.LinearClamp, DepthStencilState.None, this.GraphicsDevice.RasterizerState, this.Effect);
}

this.Draw(gameTime);

if (this.Effect != null)
{
    this.SpriteBatch.End();
    this.SpriteBatch.Begin(SpriteSortMode.Deferred, BlendState.AlphaBlend, SamplerState.LinearClamp, DepthStencilState.None, this.GraphicsDevice.RasterizerState, null);
}
```

**Comment by Zoldike** [comment]
Log:
```
if (linked == 0)
{
    var error = GL.GetErrorCode();
    var log = GL.GetProgramInfoLog(program);

    [...]

    throw new InvalidOperationException("Unable to link effect program: {" + error.ToString() + ", " + log + "}");
}
```

Error message (type 1):
>java.lang.Throwable
Msg: Unable to link effect program: {InvalidValue, No vertex shader attached.} # at Microsoft.Xna.Framework.Graphics.ShaderProgramCache.Link (Microsoft.Xna.Framework.Graphics.Shader vertexShader, Microsoft.Xna.Framework.Graphics.Shader pixelShader) [0x00000]

Error message (type 2):
>java.lang.Throwable
Msg: Unable to link effect program: {InvalidValue, L0100 A program cannot be linked unless there are any shaders attached to it # } # at Microsoft.Xna.Framework.Graphics.ShaderProgramCache.Link (Microsoft.Xna.Framework.Graphics.Shader vertexShader, Microsoft

Error message (type 3):
>java.lang.Throwable
Msg: Unable to link effect program: {InvalidValue, } # at Microsoft.Xna.Framework.Graphics.ShaderProgramCache.Link (Microsoft.Xna.Framework.Graphics.Shader vertexShader, Microsoft.Xna.Framework.Graphics.Shader pixelShader) [0x00000] in <filename unknown>:0

**Comment by SimonDarksideJ** [comment]
The 2D shaders tutorial makes use of SpriteBatch shaders without issue and I cannot replicate this issue on the latest 3.8.4.1 release, so I am closing this issue. Please reopen if it is still and issue
